### ISSUE TITLE
Device orientation vs. compass issues

### ISSUE THREAD

**Issue by andree182** [mvglasow/satstat#76]
During testing of #75 I found out that the orientation provided by Android is not too good - particularly, it the documentation says it should only be used when roll is small.

Once you rotate the device upside-down, the north heading is inverted on the compass (in a bad way). Also near the maximum roll, it behaves weirdly.

On the other hand, this seems to be common "behavior" among many apps, so perhaps this issue can be marked as won't-fix. I will probably later fix the leveller tool to use accelerometer, rather than the orientation data.

**Comment by mvglasow** [comment]
Yes, the magnetic compass in Android phones is pretty flimsy. It easily gets distracted by magnetic fields, and the sensors can be outright unusable in some devices. I once had the idea of using the compass for navigation, but I quickly abandoned that and started using GPS bearing and accelerometer data for extrapolation. While still not perfect, it is way better than the magnetic compass (to the point that I'm not even considering the compass for correction).

By upside-down, do you mean with the screen facing down? Admittedly, that's not a use case I've given a lot of thought until now. Running a brief test, I can see what you mean. In theory, we could invert the heading if the screen is facing down, but this would have another side effect: take a device with the screen facing up and tilt it 180 degrees around the X axis (so the screen is facing down in the end). As the screen goes beyond a perfectly vertical orientation, the heading would suddenly flip. I guess that would be undesirable if the device were mounted vertically in a cradle.

As for the leveller tool, that fix would be very much appreciated! I'll still hold the merge request back for a while, then.

**Comment by dlenski** [comment]
> Yes, the magnetic compass in Android phones is pretty flimsy. It easily gets distracted by magnetic fields, and the sensors can be outright unusable in some devices. I once had the idea of using the compass for navigation, but I quickly abandoned that and started using GPS bearing and accelerometer data for extrapolation. While still not perfect, it is way better than the magnetic compass (to the point that I'm not even considering the compass for correction).

Interesting. Perhaps it's just my imagination (or a function of a better GPS IC), but I have the impression that Google Maps uses the compass to improve turn-by-turn navigation.

* When walking through twisting and turning streets at low speed, GPS bearing can be unreliable. Tracing compass bearing can be a "second source" to verify which direction the user is facing/moving.
* When acquiring/reacquiring a GPS signal while stopped, compass can verify which direction a vehicle is about to start moving in, based ln the assumption that it'll move in a direction parallel to earth's surface and opposite the device screen.

**Comment by mvglasow** [comment]
Off-topic, but: Yes and Yes (though I don’t know about Google Maps).

* Real pros additionally use the gyroscope (i.e.rotation around the gravity axis) to compensate for effects from stray magnetic fields.
* The simplest way to obtain the gravity axis is the accelerometer, but this is influenced by acceleration (including centrifugal force when turning). That can be corrected by:
  * Low-pass filtering accelerometer readings
  * Measuring difference between scalar acceleration and standard gravity, and taking that as a quality indicator
  * Using the rotation vector from the gyroscope (assuming any rotations of the device are around the gravity axis—works well when the device is in a car cradle, less so when held in the user’s hand)

All of that is not quite trivial. I built something like that a while back, but bearing is still far from reliable.
